Add integration tests to api and schema fields camelCased

emilcheva opened the pull request:
Nock and jest configuration added + schema fields resolving to snake case dynamically

emilcheva committed:
Add integration tests to api and schema fields camelCased

Nock and jest configuration added + schema fields resolving to snake case dynamically

steliyan commented on package.json:
```diff
@@ -4,15 +4,25 @@
     "apollo-datasource-rest": "^3.4.0",
     "apollo-server": "^3.5.0",
-    "graphql": "^15.7.2"
+    "apollo-server-testing": "^2.25.3",
```
This should be a dev dependency.

steliyan commented on package.json:
```diff
@@ -4,15 +4,25 @@
-    "graphql": "^15.7.2"
+    "apollo-server-testing": "^2.25.3",
+    "graphql": "^15.7.2",
+    "jest": "^27.4.7"
```
This should be a dev dependency.

steliyan commented on server/index.js:
```diff
@@ -3,7 +3,18 @@ const typeDefs = require('./schema');
 const resolvers = require('./resolvers');
 const catAPI = require('./datasources/cat-api');
 
+const normalizeCamelToSnakeCase = (string) => (
+  words(toString(string).replace(/['\u2019]/g, '')).reduce((result, word, index) => (
+    result + (index ? '_' : '') + word.toLowerCase()
+  ), '')
+)
```
Oh, wow. :D

What happens if we have `something__like___that` or `__that`? A couple of unit tests covering different scenarios would be great.

steliyan commented on server/index.js:
```diff
@@ -3,7 +3,18 @@ const typeDefs = require('./schema');
+}
+
 const server = new ApolloServer({
+  fieldResolver: snakeCaseFieldResolver,
```
To be honest, I wouldn't do this here. What would happen if you consume another 3rd party service and it uses `UPPERCASE` fields instead of `snake_case`?

steliyan commented on server/test-utils/apollo-server.js:
```diff
@@ -0,0 +1,22 @@
+const resolvers = require('../resolvers');
+const typeDefs = require('../schema');
+
+const createApolloTestServer = () => {
+  const server = new ApolloServer({
+    typeDefs,
+    resolvers,
+    dataSources: () => {
+        return {
+          catAPI: new catAPI()
+        };
+      }
+  });
+  return server;
+};
```
How do the tests pass if we don't have the `fieldResolver` here?! 🤔

steliyan commented on server/test-utils/apollo-server.js:
```diff
@@ -0,0 +1,22 @@
+  const server = new ApolloServer({
+    typeDefs,
+    resolvers,
+    dataSources: () => {
+        return {
+          catAPI: new catAPI()
+        };
+      }
+  });
```
NIT: You can omit the `return` statement and use `({ catAPI: new catAPI() })`.

steliyan commented on server/test-utils/apollo-server.js:
```diff
@@ -0,0 +1,22 @@
+const { createTestClient } = require('apollo-server-testing');
+const catAPI = require('../datasources/cat-api');
+const {ApolloServer} = require('apollo-server');
```
Suggestion: Next PR ESLint and/or CI. :p

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -0,0 +1,76 @@
+    });
+
+    expect(response.errors).toBeUndefined();
+    expect(response).toMatchSnapshot();
```
Maybe `response.data.` makes more sense as you're already asserting the `response.errors`. Also, there is stuff that we dont' care about - headers, extensions, etc.

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -0,0 +1,76 @@
+import breed from "../mocks/breed.json";
+
+describe("Test Cat Wiki API", () => {
+  const baseURL = "https://api.thecatapi.com/v1";
```
NIT: You can use `BASE_URL` casing, similar to the `GET_BREEDS`.

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -0,0 +1,76 @@
+import nock from "nock";
+import { gql } from "apollo-server";
+import { query } from "../test-utils/apollo-server";
+import breed from "../mocks/breed.json";
```
Although there is a single breed, actually an array is stored. So maybe breedS is more appropriate here.

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -0,0 +1,76 @@
+    expect(response).toMatchSnapshot();
+  });
+
+  it("should throw error if no breed name is given", async () => {
```
Actually, a breed name is given, but there is no match. Because the mock data contains `persian`, maybe something that really doesn't exist should be specified, e.g. - `maine coon`.

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -0,0 +1,76 @@
+      query: GET_BREEDS_BY_NAME
+    });
+
+    expect(response.data).toBeUndefined();
```
Also, errors should be undefined as well, even though there is no match.

emilcheva committed:
Resolve PR comments

Change to json name, tests and location of package.json dependencies

steliyan commented on package.json:
```diff
@@ -13,17 +13,17 @@
-    "nodemon": "^2.0.4"
+    "nodemon": "^2.0.4",
+    "apollo-server-testing": "^2.25.3",
+    "graphql": "^15.7.2",
```
That's not a dev dependency.

steliyan commented on server/__tests__/apollo-server.test.js:
```diff
@@ -37,40 +37,56 @@ describe("Test Cat Wiki API", () => {
+    nock(BASE_URL)
+      .get("/breeds/search")
+      .query({ q: "" })
+      .reply(200);
```
I don't think that is needed as this should throw a GQL error and not hit the API at all.

emilcheva committed:
Add custom schema directive for snake_case

Lodash snakeCase is used to transform all schema fields from camelCase to SnakeCase ( when @snakeCase is explicitly set )

steliyan commented on server/resolvers.js:
```diff
@@ -6,12 +6,13 @@ const resolvers = {
     // get a single module by ID, for the module detail page
     getBreedsByName: (_, { breedName }, { dataSources }) => {
-      return dataSources.catAPI.getBreedsByName(breedName);
+      console.log(breedName)
```
Leftover?

steliyan commented on server/schema.js:
```diff
@@ -1,36 +1,79 @@
 const { gql } = require("apollo-server");
+const _ = require("lodash");
```
Not that it matters here, but generally, we import only the things we use from `lodash`, so the rest of the package can be tree-shaked.

`const snakeCase = require('lodash/snakeCase');`

steliyan commented on server/schema.js:
```diff
@@ -1,36 +1,79 @@
-type BreedImage{
-  id: ID!
-  url: String!
+// This function takes in a schema and adds snake-casing logic
+// to every resolver for an object field that has a directive with
+// the specified name (we're using `snakeCase`)
+function snakeCaseDirectiveTransformer(schema, directiveName) {
+  return mapSchema(schema, {
+    // Executes once for each object field in the schema
+    [MapperKind.OBJECT_FIELD]: (fieldConfig) => {
+      // Check whether this field has the specified directive
+      const snakeCaseDirective = getDirective(
+        schema,
+        fieldConfig,
+        directiveName
+      )?.[0];
+
+      if (snakeCaseDirective) {
+        // Get this field's original resolver
+        const { resolve = defaultFieldResolver } = fieldConfig;
+
+        fieldConfig.resolve = (source, args, context, info) => {
+          return source[_.snakeCase(info.fieldName)]
+        }
+
+        return fieldConfig;
+      }
+    },
+  });
 }
```
That seems too smart. 🤔 

Also, it's a bit weird from a GraphQL perspective. There is a `snake_case` directive defined in the GQL schema which actually is used to map data taken from a 3rd party API and match it to the schema itself.

If the 3rd party is changed (or starts provided a `camelCase`?), would you change the GQL schema?

emilcheva committed:
Add formatting and linting and field names mapping in Breed resolver

Eslint and Prettier configured + converting field names to snake_case without calling the default field resolver

emilcheva committed:
Revert "Add formatting and linting and field names mapping in Breed resolver"

This reverts commit fa5a538a141f21aff08e6044fc064ddf217e6597.

emilcheva committed:
Add camelCase to snake_case field names mapping in Breed resolver

Needed as the API returns snake case fields and GraphQL field names should be camelCase following the convention

steliyan commented on server/resolvers.js:
```diff
@@ -6,13 +6,21 @@ const resolvers = {
     },
   },
   Breed: {
+    // camelCase to snake_case mapping
```
Redundant.

steliyan commented on server/resolvers.js:
```diff
@@ -6,13 +6,21 @@ const resolvers = {
-    breedImage : ({ id }, _, { dataSources }) => {
-      return dataSources.catAPI.getImagesOfBreed(id, limit = 9);
+    breedImage: ({ id }, _, { dataSources }) => {
+      return dataSources.catAPI.getImagesOfBreed(id, (limit = 9));
```
Hm, why the parenthesis on `(limit = 9)`?

emilcheva committed:
If not provided limit will be set to 9 in the cat-API

Revert "Remove unneeded limit arg from breedImage resolver"

This reverts commit ce92a4fa35790bd78bec666190591955942e32a4.

Limit the breed images number in the cat-api

Limit should not be passed as resolver arg

emilcheva committed:
Fix for missing node modules in package.json

Rebase merge confilcts issue resolution
